doc/SONiC_VRRP_HLD.md

```diff
@@ -0,0 +1,582 @@
+
+3. VRRP owner - if a VRRP instance's virtual IP address is same as the real interface IP address. VRRP owner has special privileges (priority is 255) and can pre-empt any other router acting as master. This usually causes unwanted temporary network disruption after the non-owner master had stabilized
+  after the failover from owner.
+
+4. A non-owner master can not accept packets destined to the virtual IP address. Only the owner can
+  accept and respond to such packets.
+
+5. Two virtual routers with same VRIDs but on different IP subnets can not co-exist on the same
+  VLAN because of the virtual MAC address clash.
+
+  VRRP implementation in SONIC completely addresses limitation #1 by introducing uplink interface tracking feature.
+
+### 3.2.1    Pre-emption
+
+A VRRP Master node can be pre-empted by another VRRP node that has higher priority. Pre-emption is enabled by default. CLI is available for user to disable pre-emption on per VRRP instance. Mastership switchover causes unnecessary temporary network disruption.
+
+### 3.2.2 Virtual Router Owner
+
+VRRP instance whose virtual IP address (VIP) is same as real interface address is called the owner of virtual router and has the highest priority (255). VRRP owner is supported in SONIC's VRRP implementation. VRRP owner node will pre-empt the active master even when pre-emption is disabled for that particular VRRP instance.
+
+### 3.2.3 Virtual MAC Address
+
+Following virtual MAC address is used by the protocol (as per RFC 5798)
+
+IPv4 case: **00-00-5e-00-01-{vrid}**
+
+where, vrid is user configured 1-byte virtual router identifier. VRID has interface scope; that is, VRID has to be unique among the VRRP instances on an interface. However, same VRID can be used for two  or more virtual router instances across different interfaces.
+
+### 3.2.4    VRRP Advertisement Frame
+
+VRRP control packets have IP protocol type as 112 (reserved for VRRP), and are sent to VRRP multicast address 224.0.0.18. Source MAC in VRRP control packets is virtual MAC address and source IP is interface IP.                                                               
+
+### 3.2.5    ARP Request Handling 
+
+Only master responds to the ARP requests for virtual IP address. In ARP replies sent by master, the source MAC in Ethernet header and ARP payload is virtual MAC address.
+
+### 3.2.6   Uplink Interface Tracking
+
+Interfaces other than the VRRP instance interface can be tracked for up/down events. When interface-tracking is enabled in the VRRP instance configuration, the tracked interface's operational status will be monitored. When a interface operational down event is detected on a tracked-interface, the track-priority/weight is subtracted from the current router’s priority value. Similarly, when interface operational up event is detected on the tracked-interface, the track-priority/weight is added to the router’s current priority value.
```

Comment:
How does the uplink tracking works? for instance let's say there are more than 8 uplink interfaces how do we does it effects on mastership?

Reply:
Uplink tracking affects the overall priority for VRRP router. If the tracked interface is in 'UP' state, then the VRRP priority for the router is increased by the weight associated with the tracked interface.

Currently only 8 interfaces can be tracked per VRRP instance. Please let us know if there is a real use case of tracking more than 8 interfaces per VRRP instance.

Note multiple VRRP instances could be configured on an VRRP interface.